### ISSUE TITLE
td-agent-bit is not capturing channels information properly in Windows

### ISSUE THREAD

**Issue by yeruvass** [fluent/fluent-bit#1949]
@fujimotos  and team,

I have installed td-agent-bit on windows server and trying to fetch security channel information.

I can see security event from **windows**
<img width="660" alt="windows_security_event" src="https://user-images.githubusercontent.com/5181193/74604811-dd0e0e80-50da-11ea-9640-7a082198c3d9.png">

Same thing Pushing to **ELK** but getting below output. 

```json
{
  "_index": "abc-windows-2020.02.16",
  "_type": "_doc",
  "_id": "opT1TXAB6zoiDu5VjIgh",
  "_version": 1,
  "_score": null,
  "_source": {
    "StringInserts": [
      "S-1-5-18",
      "hostname$",
      "UATACC",
      "0x3e7",
      "{00000000-0000-0000-0000-000000000000}",
      "SYSTEM",
      "NT AUTHORITY",
      "{00000000-0000-0000-0000-000000000000}",
      "localhost",
      "localhost",
      "0x27c",
      "C:\\Windows\\System32\\services.exe",
      "-",
      "-"
    ],
    "TimeGenerated": 1581855900,
    "SourceName": "Microsoft-Windows-Security-Auditing",
    "hostname": "hostname",
    "@timestamp": "2020-02-16T12:25:03.969Z",
    "ComputerName": "computer name ",
    "EventID": 4648,
    "Sid": "",
    "@version": "1",
    "EventType": 8,
    "Data": "",
    "_@timestamp": 1581855903.380309,
    "entity": "abc-windows",
    "type": "abc-windows",
    "RecordNumber": 2318131,
    "TimeWritten": 1581855900,
    "EventCategory": 12544
  },
  "fields": {
    "@timestamp": [
      "2020-02-16T12:25:03.969Z"
    ]
  },
  "highlight": {
    "hostname": [
      "@kibana-highlighted-field@ABC@/kibana-highlighted-field@"
    ],
    "StringInserts": [
      "@kibana-highlighted-field@ABC@/kibana-highlighted-field@$"
    ],
    "hostname.keyword": [
      "@kibana-highlighted-field@ABC@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1581855903969
  ]
}
```

Looks like data is not coming properly from td-agent-bit to ELK.
I'm getting below warning in td-agent console. Is any issues with below warn logs are not parsing properly. 
```
**[ warn] [pack] invalid UTF-8 bytes, skipping**
```

**td-agent-bit.conf**

```ini
[SERVICE]
   Flush        5

[INPUT]
   Name    winlog
   Tag abc-windows-events
   Channels     Application,Windows PowerShell,Security,System
   Interval_Sec 1
   DB           winlog.sqlite

[FILTER]
   Name record_modifier
   Match abc-windows-*
   Record entity abc-windows
   Record hostname ${HOSTNAME}

[OUTPUT]
   Name kafka
   Match abc-windows-*
   Brokers ip1:9300,ip2:9300,ip3:9300
   Topics enbd-windows
   rdkafka.ssl.ca.location C:/Program Files/td-agent-bit/certs/certi.crt
   rdkafka.sasl.username test
   rdkafka.sasl.password testpassword
   rdkafka.compression.codec gzip
   rdkafka.request.required.acks 1
   rdkafka.log.connection.close false
   rdkafka.security.protocol SASL_SSL
   rdkafka.sasl.mechanisms PLAIN
```

I have installed fluentd agent on windows and results are coming properly.  

Please let me know is anything I have given wrong td-agent-bit.conf configuration.

**Comment by fujimotos** [comment]
That's probably because td-agent-bit is running as a normal user.

"Security" is a special stream that is not listenable if the process is
not running with an administrator privilege.

**Comment by yeruvass** [comment]
@fujimotos , I'm running td-agent-bit with an administrator privileges. I can read security logs but not parsing properly.

**Comment by fujimotos** [comment]
> I'm running td-agent-bit with an administrator privileges. I can read security logs but not parsing properly.

Hmm. This seems to be an encoding issue.

`in_winlog` currently doesn't normalize  the data encoding to UTF-8 (just serializes
the stream into JSON as Windows pass), so the resulting JSON might contains non-
utf8 characters on CJK environments. This seems problematic, since ELK expects
UTF-8 strings.

We need to introduce the encoding support into `in_winlog`. I'll follow up this issue
issue (probably after v1.4.0 release). Wait for me.

**Comment by fujimotos** [comment]
> I'm running td-agent-bit with an administrator privileges. I can read security logs but not parsing properly.

Hmm. This seems to be an encoding issue.

`in_winlog` currently doesn't normalize  the data encoding to UTF-8 (just serializes
the stream into JSON as Windows pass), so the resulting JSON might contains non-
utf8 characters on CJK environments. This seems to be problematic.

We need to introduce the multi-encoding support into `in_winlog`. I'll follow up this issue
issue (probably after v1.4.0 release). Wait for me.

**Comment by yeruvass** [comment]
@fujimotos , Thanks for your response. Let us know when we fix this issue

**Comment by macgahe** [comment]
@fujimotos , I did have the same issue as @yeruvass and I noticed when I was using the Fbit agent there were UTF-8 codes not "translated" properly of the agent.. I reckon you have nailed it.. and once you have a solution.. I am more than happy to give it a go..

I reckon if you look at the FluentD there is a plugin called windows_eventlog_2 which works pretty good... maybe you can have a look at this one..

**Comment by fujimotos** [comment]
@macgahe Thanks for your comment.

Indeed, I concur that the current implementation is kinda too low-level
and needs polishing up. My current plan is:

* Encode everything into UTF-8 explicitly
  * so that we have no encoding error any more.
* Polish up the output format so it becomes more readable & useful.
  * I'm planning to emit proper log messags using FormatMessage().

I'll update this ticket as my implementation proceeds, so wait for me.

**Comment by fujimotos** [comment]
@macgahe Thanks for your comment.

Indeed, I'm thinking that the current implementation is kinda too low-level
and needs polishing (the plugin right now just dumps information retrieved
from win32 API and does not try to normalize encodings).

My current plan is:

* Encode everything into UTF-8 explicitly
  * so that we have no encoding error any more.
* Polish up the output format so it becomes more readable & useful.
  * I'm planning to emit proper log messags using FormatMessage().

I'll update this ticket as my implementation proceeds, so wait for me.

**Comment by mshamota-nr** [comment]
Hi @fujimotos,
Has there been any progress on this issue? If so, do you have any idea of when could this be released? 
Many thanks in advance for your help!

**Comment by fujimotos** [comment]
@yeruvass @macgahe @mshamota-nr I updated the implementation of
`in_winlog` in #2322, so that it does not emit the UTF-8 errors.

A test build that contains the fix is available on https://github.com/fluent/fluent-bit/issues/960#issuecomment-653461701

This fix will be included in the next major version (1.5.x). I would very
appreciate if you can tell me how well it works on your environment.

Addendum: I tested the 64-bit build on my Windows Server 2019 with >10000
event log records. As far as I checked, every record was processed as expected.

**Comment by edsiper** [comment]
Fixed in #2322
